File: src/Easy/HashTable/findWords.java
package Easy.HashTable;


/**
 * 给定一个单词列表，只返回可以使用在键盘同一行的字母打印出来的单词。键盘如下图所示。
 *
 *  
 *
 *
 *
 *  
 *
 * 示例：
 *
 * 输入: ["Hello", "Alaska", "Dad", "Peace"]
 * 输出: ["Alaska", "Dad"]
 *  
 *
 * 注意：
 *
 * 你可以重复使用键盘上同一字符。
 * 你可以假设输入的字符串将只包含字母。
 **/

import java.util.ArrayList;
import java.util.List;

/**
 * @author 马世臣
 * @// TODO: 2020/2/2 500. 键盘行 */

public class findWords {

    public String[] findWords(String[] words) {
        String l1="qwertyuiop";
        String l2="asdfghjkl";
        String l3="zxcvbnm";
        List<String> strings=new ArrayList<>();
        for (String ss:words){
            String s=ss.toLowerCase();
            String l;
            char first=s.charAt(0);
            if(l1.indexOf(first)!=-1){
                l=l1;
            }else if(l2.indexOf(first)!=-1){
                l=l2;
            }else {
                l=l3;
            }
            boolean flag=true;
            for (char ch:s.toCharArray()){
                if(l.indexOf(ch)==-1){
                    flag=false;
                    break;
                }
            }
            if(flag){
                strings.add(ss);
            }
        }
        String[] strings1=new String[strings.size()];
        for (int i=0;i<strings.size();i++){
            strings1[i]=strings.get(i);
        }
        return strings1;
    }

    public static void main(String[] args) {

    }
}

File: src/Easy/HashTable/findWords.js
/**
 * @param {string[]} words
 * @return {string[]}
 */
var findWords = function (words) {
    let bucket1 = new Set();
    let bucket2 = new Set();
    let bucket3 = new Set();
    "qwertyuiop".split("").forEach(c => bucket1.add(c));
    "asdfghjkl".split("").forEach(c => bucket2.add(c));
    "zxcvbnm".split("").forEach(c => bucket3.add(c));
    let res = [];
    words.forEach(temp => {
        let word = temp.toLowerCase();
        let c = word[0];
        let flag = true;
        if (bucket1.has(c)) {
            for (let i = 1; i < word.length; i++) {
                if (!bucket1.has(word[i])) {
                    flag = false;
                    break;
                }
            }
            if (flag) {
                res.push(temp);
            }
        } else if (bucket2.has(c)) {
            for (let i = 1; i < word.length; i++) {
                if (!bucket2.has(word[i])) {
                    flag = false;
                    break;
                }
            }
            if (flag) {
                res.push(temp);
            }
        } else {
            for (let i = 1; i < word.length; i++) {
                if (!bucket3.has(word[i])) {
                    flag = false;
                    break;
                }
            }
            if (flag) {
                res.push(temp);
            }
        }
    })
    return res;
};


/**
 * @param {string[]} words
 * @return {string[]}
 */
var findWords2 = function (words) {
    return words.filter(item => {
        const wordArr = [...new Set(item.toLocaleLowerCase())]
        return [
            wordArr.every(item => 'qwertyuiop'.indexOf(item) > -1),
            wordArr.every(item => 'asdfghjkl'.indexOf(item) > -1),
            wordArr.every(item => 'zxcvbnm'.indexOf(item) > -1)
        ].some(item => item)
    })
};


// console.log([...'adsdf'].every(item=>'asdfghjkl'.indexOf(item) > -1))
// console.log([1, 2, 3].some(item => item > 2))
// console.log(findWords2(["adsdf","sfd"]))


// console.log(Array.from({length: 10}, () => 1))